### ISSUE TITLE
Senseless comparison v2 (with non strict comparison)

### ISSUE THREAD

**Issue by voku** [kalessil/phpinspectionsea#1239]
Description:
--
I know it's better to use strict comparison everywhere, but that is not so easy in legacy code, so one small step forward is maybe to reduce the usage of this kind of comparisons.

Example
--
```php
if (
	$foo == 0.00000
	||
	$foo == 0
	||
	$foo == null
	||
	$foo == ''
	||
	!$foo
) {
    // ...
}
```

reduced code with the same behaviors ...

```php
if (!$foo) {
    // ...
}
```

**Comment by orklah** [comment]
Just to be sure about what you're suggesting:

You want to add an inspection that will check for every !$foo(or $foo == false) in if statements and warn that every other loose comparison with a falsy value in the same if() is redundant?

If so, there are a few other cases to consider:
$foo == []
$foo == "0"

But there is at least one edge case:
https://3v4l.org/WPXkf
This code shows that $foo == '' and !$foo are not always equivalent with objects so it would be best to avoid the inspection if the variable could be an object and the comparison is with strings

We may consider nested constructs too:
```php
if($foo == "0"){
    if($foo == false){} // always true
    ...
    if($foo == true){} // always false
}
```
and take into account the &&/|| in the if()
```php
if($foo && $foo == "0"){} // always false
if($foo && $foo != "0"){} // second statement is redundant
```

**Comment by voku** [comment]
Yep, I think we should skip objects and arrays, because of edge cases.

**Comment by voku** [comment]
and some more edge cases like this :yum: 

```php
<?php

$a = "0.00000";

if($a == null) { // noop
    echo "float as string seems not be a null";
}
if(!$a) { // noop
    echo "float as string seems not to work here";
}
if($a == 0) { // yep
    echo "float as string seems not be a zero";
}
```

https://3v4l.org/Pjp6I

**Comment by voku** [comment]
and more ... :face_with_thermometer: 

```php
<?php

$a = [];

if ($a == null) { // yep
    echo "empty array seems to be null\n";
}
if (!$a) { // yep
    echo "empty array seems to work here\n";
}
if ($a == 0) { // noop
    echo "empty array is not zero\n";
}
```

https://3v4l.org/3TcaD

**Comment by orklah** [comment]
Yes, I didn't include the "0.0" in the list because it's not a falsy value (even if it looks like one !).

For [] == 0, it's an issue, we'll have to drop cases where the variable can be an array to avoid false positives.

To summarize, the list is:
0.0
0
null
''
"0"

who are redundant when loosely compared (==) to a variable who has already been compared to false.
The inverse should be good to, if we loosely compare with (!=) with one one thoses values, it's redundant if the variable has already been compared to true

**Comment by kalessil** [comment]
The original code case is getting reported properly now.
Tracking the if cases currently is not feasible, perhaps I can deliver this in later versions (current IDE and the plugin APIs limitation).

**Comment by orklah** [comment]
> The original code case is getting reported properly now.
> Tracking the if cases currently is not feasible, perhaps I can deliver this in later versions (current IDE and the plugin APIs limitation).

Is there a Youtrack issue to ask for a more detailed API? Perhaps we could try to vote for it?

**Comment by kalessil** [comment]
@orklah: nope, I'm in touch PhpStorm folks so at some point the APIs will materialize.
